research/engine/pegasus/pegasus/src/operator/concise/limit.rs

```diff
@@ -60,3 +66,230 @@ impl<D: Data> Limit<D> for Stream<D> {
+            move |input, output| {
+                let cmp_clone = share_cmp.cmp.clone();
+                input.for_each_batch(|dataset| {
+                    if let Some(min_value) = dataset.drain().min_by(|x, y| (*cmp_clone)(x, y)) {
+                        let new_min = if let Some(cur) = table.remove(&dataset.tag) {
+                            if (*cmp_clone)(&min_value, &cur) == Ordering::Less {
+                               min_value
+                            } else {
+                                cur
+                            }
+                        } else {
+                           min_value
+                        };
+
+                        if let Some(end) = dataset.take_end() {
+                            let mut session = output.new_session(&dataset.tag)?;
+                            session.give_last(new_min, end)?;
+                        } else {
+                            table.insert(dataset.tag.clone(), new_min);
+                        }
+
+                        return Ok(());
+                    }
+
+                    if let Some(end) = dataset.take_end() {
+                        if let Some(cur) = table.remove(&end.tag) {
+                            let mut session = output.new_session(&dataset.tag)?;
+                            session.give_last(cur, end)?;
+                        } else {
+                            output.notify_end(end)?;
+                        }
+                    }
+
+                    Ok(())
+                })
+            }
+        })
+    } else {
+        stream.unary("sort_limit_by_partition", |info| {
+            let mut table = TidyTagMap::<FixedSizeHeap<D>>::new(info.scope_level);
+            move |input, output| {
+                input.for_each_batch(|dataset| {
+                    let cmp_clone = share_cmp.cmp.clone();
+                    if !dataset.is_empty() {
+                        let heap = table.get_mut_or_else(&dataset.tag, || { FixedSizeHeap::new(size as usize, cmp_clone)});
+                        for d in dataset.drain() {
+                            heap.push(d);
+                        }
+                    }
+
+                    if dataset.is_last() {
+                        let mut session = output.new_session(&dataset.tag)?;
+                        if let Some(mut heap) = table.remove(&dataset.tag) {
+                            if last_sort {
+                                heap.sort();
+                            }
+                            let im_heap = heap;
+                            session.give_iterator(im_heap.into_iter())?;
+                        }
+                    }
+
+                    Ok(())
+                })
+            }
+        })
+    }
+}
+
+type Cmp<D> = Rc<dyn Fn(&D, &D) -> Ordering + Send + 'static>;
+
+struct ShadeCmp<C> {
+    pub cmp: Rc<C>,
+}
+
+unsafe impl<C: Send> Send for ShadeCmp<C> {}
+
+struct FixedSizeHeap<D> {
```

Comment:
We may want to use `BinaryHeap` from `std`. Note that the standard `BinaryHeap` requires the data to be `Ord`, which may not be the case for our `D` in `sort_limit_by`. For that we may implement a structure like:
```
struct DataWithCmp<D, C> {
    data: D,
    cmp: Cmp<C>
}

impl Ord for DataWithCmp ... 
```